include/VectorGraph.h

```diff
@@ -0,0 +1,739 @@
+		bool m_effectLog;
+		bool m_effectSine;
+		bool m_effectClampLower;
+		bool m_effectClampUpper;
+
+		// stores m_automationModel->value(), used in updating
+		float m_bufferedAutomationValue;
+		// automation: connecting to floatmodels, -1 when it isn't conntected'
+		int m_automationModel;
+	};
+	// deletes the point's automation model
+	// if modelLocation == point location
+	void delAutomationModel(unsigned int modelLocation, bool callDataChanged);
+	// swapping values, "slide" moves the values (between) once left or right
+	// handle m_isFixedEndPoints when using this
+	void swap(unsigned int pointLocationA, unsigned int pointLocationB, bool slide);
+	// returns the curve value at a given x coord, does clamp
+	float processCurve(float yBefore, float yAfter, float curve, float xIn);
+	// applys the effect on a given value, does clamp
+	float processEffect(unsigned int pointLocation, float attribValue, unsigned int attribLocation, float effectValue);
+	// returns a VectorGraphPoint with modified attributes, does clamp
+	float processAutomation(unsigned int pointLocation, float attribValue, unsigned int attribLocation);
+
+	// line effects / types, m_type is used for this
+	// valA: amp, valB: freq, fadeInStartLoc: from what xIn value should the line type fade out
+	//float processLineTypeSine(float xIn, float valA, float valB, float fadeInStartLoc);
+	std::vector<float> processLineTypeArraySine(std::vector<float>* xArray, unsigned int startLoc, unsigned int endLoc,
+		float valA, float valB, float fadeInStartLoc);
+	// curve: phase
+	//float processLineTypeSineB(float xIn, float valA, float valB, float curve, float fadeInStartLoc);
+	std::vector<float> processLineTypeArraySineB(std::vector<float>* xArray, unsigned int startLoc, unsigned int endLoc,
+		float valA, float valB, float curve, float fadeInStartLoc);
+	// valA: amp, valB: x coord, curve: width
+	//float processLineTypePeak(float xIn, float valA, float valB, float curve, float fadeInStartLoc);
+	std::vector<float> processLineTypeArrayPeak(std::vector<float>* xArray, unsigned int startLoc, unsigned int endLoc,
+		float valA, float valB, float curve, float fadeInStartLoc);
+	// y: calculate steps from, valA: y count, valB: curve
+	//float processLineTypeSteps(float xIn, float yIn, float valA, float valBIn, float fadeInStartLoc);
+	std::vector<float> processLineTypeArraySteps(std::vector<float>* xArray, unsigned int startLoc, unsigned int endLoc,
+		std::vector<float>* yArray, float valA, float valB, float fadeInStartLoc);
+	// valA: amp, valB: random number count, curve: seed
+	//float processLineTypeRandom(float xIn, float valA, float valB, float curve, float fadeInStartLoc);
+	std::vector<float> processLineTypeArrayRandom(std::vector<float>* xArray, unsigned int startLoc, unsigned int endLoc,
+		float valA, float valB, float curve, float fadeInStartLoc);
+
+	// updating
+	// adds the points that are
+	// effected by the effector's values changing
+	// ONLY WORKS IN SORTED ARRAYS
```

Comment:
And what happens if you pass a non-sorted array?

Reply:
It will not work. But the `effector` array's `m_needsUpdating` is always sorted.